Orange/widgets/unsupervised/owtsne.py

```diff
@@ -164,79 +326,100 @@ def _add_controls_start_box(self):
-                        if np.nan_to_num(np.nanstd(self.data.X, axis=0)).sum() \
-                                == 0:
-                            error(self.Error.constant_data)
+        if self.data is None:
+            return
+
+        if len(self.data) < 2:
+            error(self.Error.not_enough_rows)
+
+        elif not self.data.domain.attributes:
+            error(self.Error.no_attributes)
+
+        elif not self.data.is_sparse():
+            if np.all(~np.isfinite(self.data.X)):
+                error(self.Error.no_valid_data)
+            else:
+                with warnings.catch_warnings():
+                    warnings.filterwarnings(
+                        "ignore", "Degrees of freedom .*", RuntimeWarning)
+                    if np.nan_to_num(np.nanstd(self.data.X, axis=0)).sum() \
+                            == 0:
+                        error(self.Error.constant_data)
 
     def get_embedding(self):
-        if self.data is None:
+        if self.tsne_embedding is None:
             self.valid_data = None
             return None
-        elif self.projection is None:
-            embedding = np.random.normal(size=(len(self.data), 2))
-        else:
-            embedding = self.projection.embedding.X
+
+        embedding = self.tsne_embedding.embedding.X
         self.valid_data = np.ones(len(embedding), dtype=bool)
         return embedding
 
     def _toggle_run(self):
-        if self.__state == OWtSNE.Running:
-            self.stop()
+        # If no data, there's nothing to do
+        if self.data is None:
+            return
+
+        # Pause task
+        if self.task is not None:
+            self.cancel()
+            self.run_button.setText("Resume")
             self.commit()
-        elif self.__state == OWtSNE.Paused:
-            self.resume()
+        # Resume task
         else:
-            self.start()
-
-    def start(self):
-        if not self.data or self.__state == OWtSNE.Running:
-            self.graph.update_coordinates()
-        elif self.__state in (OWtSNE.Finished, OWtSNE.Waiting):
-            self.__start()
-
-    def stop(self):
-        self.__state = OWtSNE.Paused
-        self.__set_update_loop(None)
-
-    def resume(self):
-        self.__set_update_loop(self.tsne_iterator)
+            self.run()
+            self.run_button.setText("Stop")
 
     def set_data(self, data: Table):
         super().set_data(data)
 
         if data is not None:
+            n_attrs = len(data.domain.attributes)
+            self.controls.pca_components.setMaximum(
+                min(_MAX_PCA_COMPONENTS, n_attrs)
+            )
+            self.controls.pca_components.setValue(
+                min(_DEFAULT_PCA_COMPONENTS, n_attrs)
+            )
+
             # PCA doesn't support normalization on sparse data, as this would
             # require centering and normalizing the matrix
             self.normalize_cbx.setDisabled(data.is_sparse())
```

Comment:
In know this is an old code but still. It should be placed in `self.enable_controls()`. Besides, in this place `data` may differ from `self.data`, because the later could be set to None in `self.check_data()`. The same goes to the code above.

Reply:
That's neat, I didn't know this function existed.